File: services/api-gateway/app/dossier.py
"""Assemblage du dossier d'un token — fonctions pures.

Séparé de ``read_api`` pour deux raisons : ce module ne touche ni la base ni
FastAPI et se teste donc sans fixture, et ``read_api`` dépasse déjà 1500 lignes.
Il ne doit jamais importer ``read_api`` en retour (cycle d'import) — d'où le
``_iso`` local plutôt qu'un import du helper homonyme.

La règle qui gouverne tout ce fichier : une valeur non mesurée est ``None`` ou
une clé absente, jamais un ``0``. Le scoring v2 renormalise sur le poids des
axes *présents*, donc un axe absent est exclu du calcul ; le rapporter à 0.0 le
transformerait en mesure au pire, ce qui déplace le score vers le bas sans que
rien ne le signale.
"""

from __future__ import annotations

from datetime import datetime
from typing import Any

#: Les sept axes de decision-engine/app/scoring.py::WEIGHTS, dans l'ordre
#: d'affichage. Dupliqués ici et non importés : api-gateway ne dépend pas du
#: decision-engine, et cette liste ne bouge que lors d'un changement de modèle
#: de scoring, qui touchera de toute façon les deux fichiers.
AXIS_KEYS: tuple[str, ...] = (
    "volume_growth",
    "social_score",
    "news_score",
    "market_trend",
    "liquidity_score",
    "positioning",
    "fundamentals",
)


def _iso(v: Any) -> str | None:
    if v is None:
        return None
    if isinstance(v, datetime):
        return v.isoformat()
    return str(v)


def build_score(journal: Any | None) -> dict:
    """Décomposition par axe du dernier score connu pour un symbole.

    ``axes`` ne contient que les axes **mesurés**. L'absence d'une clé est
    l'information : elle dit « non mesuré », pas « nul ».
    """
    if journal is None:
        return {
            "value": None,
            "confidence": None,
            "axes": {},
            "axes_total": len(AXIS_KEYS),
            "dominant_factor": None,
            "dominant_factor_share": None,
            "computed_at": None,
        }

    factors = journal.factors or {}
    return {
        "value": journal.score,
        "confidence": journal.confidence,
        # `is not None` et non un test de vérité : un axe mesuré à 0.0 est une
        # mesure et doit être conservé.
        "axes": {k: float(factors[k]) for k in AXIS_KEYS if factors.get(k) is not None},
        "axes_total": len(AXIS_KEYS),
        "dominant_factor": journal.dominant_factor,
        "dominant_factor_share": journal.dominant_factor_share,
        "computed_at": _iso(journal.time),
    }

File: frontend/src/lib/types/dossier.ts
import type { NewsItem, Position, Trade, WorkerDecision } from './domain';

/**
 * Les sept axes de decision-engine/app/scoring.py::WEIGHTS, dans l'ordre
 * d'affichage du drawer.
 *
 * Troisième copie indépendante de cette liste, avec
 * `decision-engine/app/scoring.py::WEIGHTS` et
 * `services/api-gateway/app/dossier.py::AXIS_KEYS`. Rien ne vérifie qu'elles
 * restent alignées : un axe ajouté au scoring sans être ajouté ici
 * n'apparaîtrait jamais à l'écran, sans erreur ni test rouge.
 */
export const SCORE_AXES = [
  'volume_growth',
  'social_score',
  'news_score',
  'market_trend',
  'liquidity_score',
  'positioning',
  'fundamentals',
  'developer_activity',
] as const;

export type ScoreAxis = (typeof SCORE_AXES)[number];

export const AXIS_LABELS: Record<ScoreAxis, string> = {
  volume_growth: 'Volume',
  social_score: 'Social',
  news_score: 'News',
  market_trend: 'Tendance',
  liquidity_score: 'Liquidité',
  positioning: 'Positionnement',
  fundamentals: 'Fondamentaux',
  developer_activity: 'Développement',
};

export interface TokenScore {
  /**
   * Sur **0–100** — le `Decision.opportunity_score` brut, rendu tel quel à côté
   * d'un « / 100 ». Attention au voisin : `TokenDossier.decisions[].opportunity_score`
   * est sur 0–1 dans la même réponse (`map_decision` divise), et `ScoreChip`
   * remultiplie. Deux échelles pour « score d'opportunité » dans un même payload.
   */
  value: number | null;
  confidence: number | null;
  /**
   * Seuls les axes **mesurés** sont présents. Une clé absente signifie « non
   * mesuré » : le scoring renormalise sur le poids présent, donc l'axe est
   * exclu du calcul, pas noté zéro. Le type est `Partial` exprès — il force
   * l'appelant à traiter `undefined`, ce qu'un `Record` complet masquerait.
   */
  axes: Partial<Record<ScoreAxis, number>>;
  axes_total: number;
  /**
   * `true` : une décision existe, mais le poids des axes présents était sous le
   * seuil de renormalisation — le back renvoie alors `value: null` plutôt que le
   * `0` que le moteur de scoring produit dans ce cas. `false` quand aucune
   * décision n'existe : rien n'a été tenté, ce n'est pas la même chose.
   */
  insufficient_evidence: boolean;
  computed_at: string | null;
}

export interface PipelineVerdict {
  /** Vocabulaire de systems_pipeline.py::STAGE_SPECS. */
  reached_stage: string | null;
  /** `null` = aucun blocage observé. Ce n'est pas « passé » : ce peut être un
   *  signal encore en vol. L'UI doit distinguer les deux. */
  blocked_at: string | null;
  block_reason: string | null;
  /** `null` quand la seule trace est un rejet sans ligne de journal : on ignore
   *  alors si Haiku avait escaladé. Ne pas rendre `null` comme « non ». */
  escalated: boolean | null;
  sonnet_called: boolean | null;
  sonnet_validated: boolean | null;
  last_event_at: string | null;
}

export interface TokenExposure {
  open_positions: Position[];
  recent_trades: Trade[];
}

export interface TokenDossier {
  symbol: string;
  score: TokenScore;
  pipeline: PipelineVerdict;
  decisions: WorkerDecision[];
  /** News **et** social mentionnant le symbole. */
  content: NewsItem[];
  exposure: TokenExposure;
}

/**
 * Valeur d'un axe, `undefined` normalisé en `null`.
 *
 * Exister sous cette forme est le point : `axes[axis] ?? 0` est l'erreur que ce
 * projet a déjà commise sous d'autres formes, et elle ne lève rien. Passer par
 * ce helper rend l'absence explicite au site d'appel.
 */
export function axisValue(
  axes: TokenScore['axes'],
  axis: ScoreAxis,
): number | null {
  const v = axes[axis];
  return v === undefined ? null : v;
}
